In [1]:
import pandas as pd

In [25]:
# download data and put in the data/ folder. Then change below to data/cdph_data.csv
source_data = "data/CDPH Search results - 1_12_2026, 3_34_15 PM.csv"
cdph = pd.read_csv(source_data, low_memory = False)
# ORIGINAL cdph = pd.read_csv("cdph_data.csv", low_memory = False)
len(cdph)

534508

In [18]:
len(cdph[cdph["Brand"] == "Glossier"])

636

In [14]:
cdph["Ingredient Name"].unique()

array(['Titanium dioxide (CI 77891) 13463-67-7 / 1317-70-0 / 1317-80-2 / 98084-96-9',
       'CITRONELLOL ', 'LINALOOL ', 'HEXYL CINNAMAL ',
       'Limonene (1-methyl-4-prop-1-en-2-yl-cyclohexene; dl-limonene (racemic); Dipentene; (R)-p-mentha-1,8-diene; (d-limonene); (S)-p-mentha-1,8-diene; (l-limonene)) 5989-27-5 / 138-86-3 / 7705-14-8 / 5989-54-8',
       'Lilial ',
       'Aloe vera, non-decolorized (unfiltered) whole leaf extract ',
       '1,6-Octadien-3-ol, 3,7-dimethyl- (LINALOOL) 78-70-6', 'CITRAL ',
       'GERANIOL ',
       'Citronellol /+- 3,7-dimethyloct-6-en-1-ol (CITRONELLOL; (3R)-3,7-dimethyloct-6-en-1-ol; (3S)-3,7-dimethyloct-6-en-1-ol) 106-22-9 / 26489-01-0 / 1117-61-9 / 7540-51-4',
       'Benzyl benzoate 120-51-4', 'BUTYLPHENYL METHYLPROPIONAL ',
       '2-Benzylideneoctanal (HEXYL CINNAMAL) 101-86-0',
       '2-Phenoxyethyl isobutyrate 103-60-6',
       '1,3,4,6,7,8-Hexahydro-4,6,6,7,8,8-hexamethylcyclopenta-?-2-benzopyran (Hexamethylindanopyran) 1222-05-5',
    

In [9]:
len(cdph)

534508

In [24]:
cdph.head()

,Product Id,Company,Brand,Product Name,Variant,Product Discontinued Date,Product Submitted Date,Ingredient Name,Function,Unit of Measure,Concentration,Ingredient Submitted Date,Ingredient Removed Date
0,18358,Nail Alliance - Entity,Entity Nudite,Cool Pink Nail Sculpting Powder,NaN,NaN,6/24/2019,Titanium dioxide (CI 77891) 13463-67-7 / 1317-...,NaN,mg/g,5.000000,6/24/2019,NaN
1,38661,"Xtreme Color, Inc.",Hard Candy,Fierce Effects- Shadow Duo,Black-Solid,NaN,9/21/2015,Titanium dioxide (CI 77891) 13463-67-7 / 1317-...,NaN,mg/g,40.000000,9/21/2015,NaN
2,38662,"Xtreme Color, Inc.",Hard Candy,Fierce Effects- Shadow Duo,Black-Sprinkle,NaN,9/21/2015,Titanium dioxide (CI 77891) 13463-67-7 / 1317-...,NaN,mg/g,31.000000,9/21/2015,NaN
3,38667,"Xtreme Color, Inc.",Hard Candy,Fierce Effects- Shadow Duo,Brown-Solid,NaN,9/21/2015,Titanium dioxide (CI 77891) 13463-67-7 / 1317-...,NaN,mg/g,12.000000,9/21/2015,NaN
4,38665,"Xtreme Color, Inc.",Hard Candy,Fierce Effects- Shadow Duo,Brown-Solid,NaN,9/21/2015,Titanium dioxide (CI 77891) 13463-67-7 / 1317-...,NaN,mg/g,21.500000,9/21/2015,NaN


In [22]:
null = cdph[cdph['Ingredient Name'].isna()]
null.head()

,Product Id,Company,Brand,Product Name,Variant,Product Discontinued Date,Product Submitted Date,Ingredient Name,Function,Unit of Measure,Concentration,Ingredient Submitted Date,Ingredient Removed Date
1402,72818,Exclusive Fragrances & Cosmetics,10 Corso Como,10 Corso Como Massage Oil,10 Corso Como,NaN,6/29/2010,NaN,NaN,NaN,NaN,NaN,NaN
1404,72816,IPD Inc,10 Corso Como,Corso Como Bath Pearls,Corso Como,NaN,6/25/2010,NaN,NaN,NaN,NaN,NaN,NaN
10283,109738,Aloisia Beauty Products Inc.,Aloisia Beauty,Light Hydration Emulsion,NaN,NaN,6/6/2022,NaN,NaN,NaN,NaN,NaN,NaN
10284,109736,Aloisia Beauty Products Inc.,Aloisia Beauty,Renew Concentrated Serum,NaN,NaN,6/6/2022,NaN,NaN,NaN,NaN,NaN,NaN
10286,109739,Aloisia Beauty Products Inc.,Aloisia Beauty,Revitalize Eye Cream,NaN,NaN,6/6/2022,NaN,NaN,NaN,NaN,NaN,NaN


In [4]:
ts = cdph[cdph["Concentration"] == "Trade Secret"]

In [23]:
cdph["Ingredient Name"].nunique()

495

In [5]:
ts.head()

,Product Id,Company,Brand,Product Name,Variant,Product Discontinued Date,Product Submitted Date,Ingredient Name,Function,Unit of Measure,Concentration,Ingredient Submitted Date,Ingredient Removed Date
151,32263,Mary Kay Inc.,Mary Kay,Blonde Mechanical Brow Pencil,BLACK IRON OXIDE,3/15/2010,3/26/2014,Titanium dioxide (CI 77891) 13463-67-7 / 1317-...,NaN,mg/g,Trade Secret,3/26/2014,NaN
152,32264,Mary Kay Inc.,Mary Kay,Blonde Mechanical Brow Pencil,COSMETIC RUSSET IRON OXIDE,3/15/2010,3/26/2014,Titanium dioxide (CI 77891) 13463-67-7 / 1317-...,NaN,mg/g,Trade Secret,3/26/2014,NaN
153,32273,Mary Kay Inc.,Mary Kay,Brow Definer Pencil,BLACK IRON OXIDE,NaN,3/26/2014,Titanium dioxide (CI 77891) 13463-67-7 / 1317-...,NaN,mg/g,Trade Secret,3/26/2014,NaN
154,32275,Mary Kay Inc.,Mary Kay,Brow Definer Pencil,COSMETIC RUSSET IRON OXIDE,NaN,3/26/2014,Titanium dioxide (CI 77891) 13463-67-7 / 1317-...,NaN,mg/g,Trade Secret,3/26/2014,NaN
155,32274,Mary Kay Inc.,Mary Kay,Brow Definer Pencil,COSMETIC YELLOW IRON OXIDE,NaN,3/26/2014,Titanium dioxide (CI 77891) 13463-67-7 / 1317-...,NaN,mg/g,Trade Secret,3/26/2014,NaN


In [8]:
pd.set_option("display.max_rows", None)
ts_brand = ts.groupby("Brand").size().rename("count").reset_index()
ts_brand.sort_values("count", ascending = False)

,Brand,count
628,Kryolan,16411
373,Elizabeth Arden,2638
829,PERLIER,1795
188,CLE DE PEAU BEAUTE,1764
599,KIKO MILANO,1708
232,Charlotte Tilbury,1549
81,Aveda,1540
470,Granado,1521
950,SHISEIDO,1376
492,HARBOR NATURAL COSMETICS,1353


In [ ]:
cdph.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 534508 entries, 0 to 534507
Data columns (total 17 columns):
 #   Column                     Non-Null Count   Dtype 
---  ------                     --------------   ----- 
 0   Product Id                 534508 non-null  int64 
 1   Company                    534508 non-null  object
 2   Brand                      534308 non-null  object
 3   Product Name               534508 non-null  object
 4   Variant                    187813 non-null  object
 5   Product Discontinued Date  39102 non-null   object
 6   Product Submitted Date     534508 non-null  object
 7   Ingredient Name            534186 non-null  object
 8   Function                   433384 non-null  object
 9   Unit of Measure            222747 non-null  object
 10  Concentration              224495 non-null  object
 11  Ingredient Submitted Date  534186 non-null  object
 12  Ingredient Removed Date    6627 non-null    object
 13  discontinued_year          39102 non-null   

In [ ]:
cdph["Function"].unique()

array([nan, 'Fragrance', 'Conditioning/Emollient', 'Other (Specify)',
       'Preservative/Stabilizer', 'Flavor', 'Opacifying Agent',
       'Bulking/Viscosity', 'Solvent', 'Surfactant', 'Colorant',
       'Anticaking/Absorbents', 'Antioxidant', 'Moisturizing/Humectant',
       'Buffering/pH Adjusting', 'UV Inhibitor', 'Contaminant',
       'Emulsifier', 'Base'], dtype=object)

In [ ]:
cdph.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 534508 entries, 0 to 534507
Data columns (total 13 columns):
 #   Column                     Non-Null Count   Dtype 
---  ------                     --------------   ----- 
 0   Product Id                 534508 non-null  int64 
 1   Company                    534508 non-null  object
 2   Brand                      534308 non-null  object
 3   Product Name               534508 non-null  object
 4   Variant                    187813 non-null  object
 5   Product Discontinued Date  39102 non-null   object
 6   Product Submitted Date     534508 non-null  object
 7   Ingredient Name            534186 non-null  object
 8   Function                   433384 non-null  object
 9   Unit of Measure            222747 non-null  object
 10  Concentration              224495 non-null  object
 11  Ingredient Submitted Date  534186 non-null  object
 12  Ingredient Removed Date    6627 non-null    object
dtypes: int64(1), object(12)
memory usage: 53.0+ 

There have been 534,508 products reported since the beginning of the program

39,102 have been discontinued (appx 7%)

In [ ]:
cdph["discontinued_year"] = pd.to_datetime(cdph["Product Discontinued Date"]).dt.year
cdph["product_submitted_year"] = pd.to_datetime(cdph["Product Submitted Date"]).dt.year
cdph["ingredient_submitted_date"] = pd.to_datetime(cdph["Ingredient Submitted Date"]).dt.year
cdph["ingredient_removed_year"] = pd.to_datetime(cdph["Ingredient Removed Date"]).dt.year

In [ ]:
cdph["discontinued_year"] = cdph["discontinued_year"].astype("Int64")
cdph["product_submitted_year"] = cdph["product_submitted_year"].astype("Int64")
cdph["ingredient_submitted_date"] = cdph["ingredient_submitted_date"].astype("Int64")
cdph["ingredient_removed_year"] = cdph["ingredient_removed_year"].astype("Int64")

In [ ]:
cdph.head()

,Product Id,Company,Brand,Product Name,Variant,Product Discontinued Date,Product Submitted Date,Ingredient Name,Function,Unit of Measure,Concentration,Ingredient Submitted Date,Ingredient Removed Date,discontinued_year,product_submitted_year,ingredient_submitted_date,ingredient_removed_year
0,18358,Nail Alliance - Entity,Entity Nudite,Cool Pink Nail Sculpting Powder,NaN,NaN,6/24/2019,Titanium dioxide (CI 77891) 13463-67-7 / 1317-...,NaN,mg/g,5.000000,6/24/2019,NaN,<NA>,2019,2019,<NA>
1,38661,"Xtreme Color, Inc.",Hard Candy,Fierce Effects- Shadow Duo,Black-Solid,NaN,9/21/2015,Titanium dioxide (CI 77891) 13463-67-7 / 1317-...,NaN,mg/g,40.000000,9/21/2015,NaN,<NA>,2015,2015,<NA>
2,38662,"Xtreme Color, Inc.",Hard Candy,Fierce Effects- Shadow Duo,Black-Sprinkle,NaN,9/21/2015,Titanium dioxide (CI 77891) 13463-67-7 / 1317-...,NaN,mg/g,31.000000,9/21/2015,NaN,<NA>,2015,2015,<NA>
3,38667,"Xtreme Color, Inc.",Hard Candy,Fierce Effects- Shadow Duo,Brown-Solid,NaN,9/21/2015,Titanium dioxide (CI 77891) 13463-67-7 / 1317-...,NaN,mg/g,12.000000,9/21/2015,NaN,<NA>,2015,2015,<NA>
4,38665,"Xtreme Color, Inc.",Hard Candy,Fierce Effects- Shadow Duo,Brown-Solid,NaN,9/21/2015,Titanium dioxide (CI 77891) 13463-67-7 / 1317-...,NaN,mg/g,21.500000,9/21/2015,NaN,<NA>,2015,2015,<NA>


In [ ]:
no_ts = cdph[cdph["Concentration"] != "Trade Secret"]

In [ ]:
no_ts["Concentration"] = no_ts["Concentration"].astype("float")

/var/folders/p6/tqnlcwnd075_z06_lxt45pv40000gn/T/ipykernel_91911/1848226041.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  no_ts["Concentration"] = no_ts["Concentration"].astype("float")


In [ ]:
no_ts.info()

<class 'pandas.core.frame.DataFrame'>
Index: 420446 entries, 0 to 534507
Data columns (total 17 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   Product Id                 420446 non-null  int64  
 1   Company                    420446 non-null  object 
 2   Brand                      420262 non-null  object 
 3   Product Name               420446 non-null  object 
 4   Variant                    143897 non-null  object 
 5   Product Discontinued Date  36508 non-null   object 
 6   Product Submitted Date     420446 non-null  object 
 7   Ingredient Name            420124 non-null  object 
 8   Function                   322652 non-null  object 
 9   Unit of Measure            109965 non-null  object 
 10  Concentration              110433 non-null  float64
 11  Ingredient Submitted Date  420124 non-null  object 
 12  Ingredient Removed Date    5764 non-null    object 
 13  discontinued_year          36508 n

In [ ]:
no_ts["Concentration"].mean()

np.float64(22.88877946678982)

In [ ]:
no_ts.info()

<class 'pandas.core.frame.DataFrame'>
Index: 420446 entries, 0 to 534507
Data columns (total 17 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   Product Id                 420446 non-null  int64  
 1   Company                    420446 non-null  object 
 2   Brand                      420262 non-null  object 
 3   Product Name               420446 non-null  object 
 4   Variant                    143897 non-null  object 
 5   Product Discontinued Date  36508 non-null   object 
 6   Product Submitted Date     420446 non-null  object 
 7   Ingredient Name            420124 non-null  object 
 8   Function                   322652 non-null  object 
 9   Unit of Measure            109965 non-null  object 
 10  Concentration              110433 non-null  float64
 11  Ingredient Submitted Date  420124 non-null  object 
 12  Ingredient Removed Date    5764 non-null    object 
 13  discontinued_year          36508 n

In [ ]:
import altair as alt

In [ ]:
cdph.sort_values("Concentration")

In [ ]:
submissions = cdph.groupby("product_submitted_year").size().rename("product_count").reset_index()
submissions

,product_submitted_year,product_count
0,2009,28251
1,2010,13522
2,2011,4191
3,2012,3317
4,2013,5787
5,2014,7647
6,2015,6825
7,2016,4008
8,2017,7256
9,2018,9387


In [ ]:
discontinued = cdph.groupby("discontinued_year").size().rename("product_count").reset_index()
discontinued

,discontinued_year,product_count
0,2001,2
1,2004,21
2,2005,6
3,2006,16
4,2007,222
5,2008,366
6,2009,742
7,2010,1064
8,2011,1241
9,2012,975


In [ ]:
total_products = discontinued["product_count"].sum()
print(total_products)

39102


In [ ]:
import altair as alt

# Discontinued Products by Year

In [ ]:
alt.Chart(discontinued).mark_bar().encode(
    x="discontinued_year",
    y="product_count"
)

alt.Chart(...)

# Submitted Products by Year

In [ ]:
alt.Chart(submissions).mark_bar().encode(
    x="product_submitted_year",
    y="product_count"
)

alt.Chart(...)

In [ ]:
ingredients = cdph.groupby("Ingredient Name").size().rename("total_count").reset_index()
ingredients.sort_values("total_count", ascending = False)

,Ingredient Name,total_count
451,Titanium dioxide (CI 77891) 13463-67-7 / 1317-...,91510
311,Limonene (1-methyl-4-prop-1-en-2-yl-cyclohexen...,42034
10,"1,6-Octadien-3-ol, 3,7-dimethyl- (LINALOOL) 78...",29596
125,Benzyl alcohol 100-51-6,17246
128,Benzyl salicylate 118-58-1,15145
302,LINALOOL,14827
183,"Citronellol /+- 3,7-dimethyloct-6-en-1-ol (CIT...",14264
126,Benzyl benzoate 120-51-4,14030
25,"2,6-Octadien-1-ol, 3,7-dimethyl-, (2E)- (GERAN...",13247
157,CITRONELLOL,11504


In [ ]:
pd.set_option('display.max_rows', None)
brands = cdph.groupby("Brand").size().rename("total_count").reset_index()
top_50_brands = brands.sort_values("total_count", ascending = False).head(50)
top_50_brands

,Brand,total_count
551,Bath & Body Works,57104
2143,Kryolan,22538
4136,Victoria's Secret Beauty,15386
2336,Lush Handmade Cosmetics,9285
1242,Dove,4778
3448,Sephora,4735
911,Charlotte Tilbury,4330
2993,Parfums Christian Dior,4049
1346,Elizabeth Arden,3998
2349,L’OCCITANE EN PROVENCE,3981


In [ ]:
alt.Chart(top_50_brands).mark_bar().encode(
    x = "total_count",
    y = alt.Y("Brand").sort("-x")
)

alt.Chart(...)

In [ ]:
product_types = cdph.groupby("Product Category").size().rename("product_count").reset_index().sort_values("product_count", ascending=False)

KeyError: 'Product Category'

In [ ]:
pd.set_option('display.max_rows', None)
product_types.sort_values("product_count", ascending = False)
product_types

,Product Category,product_count
25,Foundations and Bases,237005
68,Perfumes/Eaux de Parfum,231647
79,Skin Moisturizers (making a cosmetic claim),79633
87,Toilet Water/Eaux de Toilette,74532
11,Body Washes and Soaps,65144
63,Other Skin Care Product,48725
41,"Lip Color - Lipsticks, Liners, and Pencils",40441
33,Hair Shampoos (making a cosmetic claim),38935
29,Hair Conditioners (rinse-out),32932
55,Other Fragrances,30501


In [ ]:
alt.Chart()

In [ ]:
thisyear=cdph[cdph["discontinued_year"] == 2025]
thisyear.head()

,Product Id,Company,Brand,Product Name,Variant,Product Discontinued Date,Product Submitted Date,Ingredient Name,Function,Unit of Measure,...,Ingredient Removed Date,UPC,Body Area,Product Category,Product Form,Intended Market,discontinued_year,product_submitted_year,ingredient_submitted_date,ingredient_removed_year
890,119884,Davines S.p.A.,[comfort zone],/ SKIN REGIMEN / cleansing cream,NaN,10/15/2025,12/30/2022,Limonene (1-methyl-4-prop-1-en-2-yl-cyclohexen...,Fragrance,NaN,...,NaN,8004608516866,Face,Skin Cleansers,Cream/Gel/Foam,General population,2025,2022,2023,<NA>
891,119884,Davines S.p.A.,[comfort zone],/ SKIN REGIMEN / cleansing cream,NaN,10/15/2025,12/30/2022,Diethyl phthalate (DEP) 84-66-2,Fragrance,NaN,...,NaN,8004608516866,Face,Skin Cleansers,Cream/Gel/Foam,General population,2025,2022,2022,<NA>
892,119884,Davines S.p.A.,[comfort zone],/ SKIN REGIMEN / cleansing cream,NaN,10/15/2025,12/30/2022,LINALOOL,Fragrance,NaN,...,NaN,8004608516866,Face,Skin Cleansers,Cream/Gel/Foam,General population,2025,2022,2023,<NA>
895,130717,Davines S.p.A.,[comfort zone],/ SKIN REGIMEN / hand cream,NaN,10/15/2025,10/24/2023,Limonene (1-methyl-4-prop-1-en-2-yl-cyclohexen...,Fragrance,NaN,...,NaN,8004608500056,Hands,Other Skin Care Product,Cream/Gel/Foam,General population,2025,2023,2023,<NA>
896,130717,Davines S.p.A.,[comfort zone],/ SKIN REGIMEN / hand cream,NaN,10/15/2025,10/24/2023,LINALOOL,Fragrance,NaN,...,NaN,8004608500056,Hands,Other Skin Care Product,Cream/Gel/Foam,General population,2025,2023,2023,<NA>


need to calculate length of time from product submitted date to product discontinued date

In [ ]:
thisyear.groupby("Product Category").size().rename("product_count").reset_index().sort_values("product_count", ascending = False)

,Product Category,product_count
40,Other Skin Care Product,2658
52,Skin Moisturizers (making a cosmetic claim),2165
58,Toilet Water/Eaux de Toilette,1894
5,Body Washes and Soaps,1680
45,Perfumes/Eaux de Parfum,1259
12,Facial Cream,894
34,Other Fragrances,526
26,"Lip Color - Lipsticks, Liners, and Pencils",346
35,Other Hair Care Product,343
48,Scrubs and Exfoliants,318


In [ ]:
thisyear.groupby("Brand").size().rename("product_count").reset_index().sort_values("product_count", ascending = False)

,Brand,product_count
85,The Body Shop International Limited,9259
84,The Body Shop,1596
41,Lush Handmade Cosmetics,788
24,Gold Bond,639
57,Neal's Yard Remedies,526
73,SERGE LUTENS,407
5,BEAUTY RENEW,263
22,Glee,185
32,ISSEY MIYAKE - FUSION D'ISSEY,162
64,PB Premiere Editions,148
